benchmarktests/target_secret_totp.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package benchmarktests
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"math/rand"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/hashicorp/vault/api"
+	vegeta "github.com/tsenart/vegeta/v12/lib"
+)
+
+const (
+	TOTPReadTestType    = "totp_read"
+	TOTPWriteTestType   = "totp_write"
+	TOTPReadTestMethod  = "GET"
+	TOTPWriteTestMethod = "POST"
+)
+
+func init() {
+	TestList[TOTPReadTestType] = func() BenchmarkBuilder { return &TOTPTest{action: "read"} }
+	TestList[TOTPWriteTestType] = func() BenchmarkBuilder { return &TOTPTest{action: "write"} }
+}
+
+type TOTPTest struct {
+	pathPrefix string
+	header     http.Header
+	config     *TOTPSecretTestConfig
+	numkeys    int
+	keysize    int
+	action     string
+	logger     hclog.Logger
+}
+
+type TOTPSecretTestConfig struct {
+	NumKeys int `hcl:"numkeys,optional"`
+	KeySize int `hcl:"keysize,optional"`
+}
+
+func (a *TOTPTest) ParseConfig(body hcl.Body) error {
+	testConfig := &struct {
+		Config *TOTPSecretTestConfig `hcl:"config,block"`
+	}{
+		Config: &TOTPSecretTestConfig{
+			KeySize: 20,
+			NumKeys: 1000,
+		},
+	}
+
+	diags := gohcl.DecodeBody(body, nil, testConfig)
+	if diags.HasErrors() {
+		return fmt.Errorf("error decoding to struct: %v", diags)
+	}
+	a.config = testConfig.Config
+
+	return nil
+}
+
+func (t *TOTPTest) read(client *api.Client) vegeta.Target {
+	secnum := int(1 + rand.Int31n(int32(t.numkeys)))
+	return vegeta.Target{
+		Method: TOTPReadTestMethod,
+		URL:    client.Address() + t.pathPrefix + "/code/" + strconv.Itoa(secnum),
+		Header: t.header,
+	}
+}
+
+func (t *TOTPTest) write(client *api.Client) vegeta.Target {
+	secnum := int(1 + rand.Int31n(int32(t.numkeys)))
+	return vegeta.Target{
+		Method: TOTPWriteTestMethod,
+		URL:    client.Address() + t.pathPrefix + "/keys/write-test-" + strconv.Itoa(secnum),
+		Header: t.header,
+		Body:   []byte(`{"name":"` + strconv.Itoa(secnum) + `","generate":true,"issuer":"Vault","account_name":"test@example.com"}`),
```

Comment:
The data being passed here should be parameterized using configuration passed in from above comment.